### PR TITLE
Dynamic sampling

### PR DESCRIPTION
**Pull request by janstenpickle**
Allows head sampler configuration to be updated at runtime

### Todo

- [x] Docs
- [x] Example
- [ ] Config file source

### CODE REVIEW

File: modules/rate-sampling/src/main/scala/io/janstenpickle/trace4cats/rate/sampling/RateSpanSampler.scala [hunk 1]
```diff
@@ -1,34 +1,54 @@
...
         }
+
+        parentContext.map(_.traceFlags.sampled).fold(sampleDecisionF) {
+          case SampleDecision.Include =>
+            if (rootSpansOnly) Applicative[F].pure(SampleDecision.Include)
+            else sampleDecisionF
+          case SampleDecision.Drop => Applicative[F].pure(SampleDecision.Drop)
+        }
+      }
+
```
Comment (janstenpickle) [inline modules/rate-sampling/src/main/scala/io/janstenpickle/trace4cats/rate/sampling/RateSpanSampler.scala:35]: I've change the functionality here to be align more with the probabilistic sampler. We may want to backport.

File: build.sbt [hunk 2]
```diff
@@ -631,6 +635,32 @@ lazy val filtering = (project in file("modules/filtering"))
   .settings(name := "trace4cats-filtering", libraryDependencies ++= Dependencies.test.map(_ % Test))
   .dependsOn(model, kernel, `exporter-stream`)
 
+lazy val `dynamic-sampling` = (project in file("modules/dynamic-sampling"))
+  .settings(publishSettings)
+  .settings(name := "trace4cats-dynamic-sampling", libraryDependencies ++= Dependencies.test.map(_ % Test))
+  .dependsOn(model, kernel, `rate-sampling`)
```
Comment (catostrophe) [inline build.sbt:641]:
```suggestion
  .dependsOn(model, kernel, `rate-sampling`, test % "test->compile")
```

Doesn't work for me without this.

File: modules/rate-sampling/src/main/scala/io/janstenpickle/trace4cats/rate/DynamicTokenBucket.scala [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+        override def request(n: Int): F[Int] = underlying.request(n)
+      },
+      for {
+        fiber <- processFiberRef.get
+        _ <- currentConfig.update { case (conf, _) => (conf, true) }
+        _ <- fiber.cancel
+      } yield ()
+    )
+}
```
Comment (catostrophe) [inline modules/rate-sampling/src/main/scala/io/janstenpickle/trace4cats/rate/DynamicTokenBucket.scala:57]:
```suggestion
package io.janstenpickle.trace4cats.rate

import cats.Applicative
import cats.effect.kernel.Ref
import cats.effect.kernel.syntax.spawn._
import cats.effect.std.Hotswap
import cats.effect.kernel.{Resource, Temporal}
import cats.syntax.flatMap._
import cats.syntax.functor._

import scala.concurrent.duration.FiniteDuration

trait DynamicTokenBucket[F[_]] extends TokenBucket[F] {
  def updateConfig(bucketSize: Int, tokenRate: FiniteDuration): F[Unit]
}

object DynamicTokenBucket {
  type CancelToken[F[_]] = F[Unit]

  def create[F[_]: Temporal](bucketSize: Int, tokenRate: FiniteDuration): Resource[F, DynamicTokenBucket[F]] =
    for {
      currentConfig <- Resource.eval(Ref.of[F, (Int, FiniteDuration)]((bucketSize, tokenRate)))
      tokens <- Resource.eval(Ref.of(bucketSize))
      (hotswap, _) <- Hotswap(TokenBucket.bucketProcess(tokens, bucketSize, tokenRate).background.void)
    } yield new DynamicTokenBucket[F] {
      private final val underlying = TokenBucket.impl[F](tokens)

      override def updateConfig(bucketSize: Int, tokenRate: FiniteDuration): F[Unit] =
        currentConfig.get.flatMap { oldConfig =>
          Applicative[F].whenA(oldConfig != (bucketSize -> tokenRate)) {
            hotswap.swap(TokenBucket.bucketProcess(tokens, bucketSize, tokenRate).background.void)
          }
        }

      override def request1: F[Boolean] = underlying.request1

      override def request(n: Int): F[Int] = underlying.request(n)
    }
}
```

File: modules/rate-sampling/src/test/scala/io/janstenpickle/trace4cats/rate/DynamicTokenBucketSpec.scala [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+        tokens2 <- bucket.request(2)
+      } yield (tokens0, tokens1, tokens2)
+    }
+
+    val result = test.unsafeToFuture()
+    ticker.ctx.tick(20.seconds)
+    result.value shouldEqual Some(Success((2, 0, 1)))
+  }
+
+  behavior.of("DynamicTokenBucket.unsafeCreate")
```
Comment (catostrophe) [inline modules/rate-sampling/src/test/scala/io/janstenpickle/trace4cats/rate/DynamicTokenBucketSpec.scala:52]: This is not needed anymore.

**Review by catostrophe — CHANGES_REQUESTED** [review]
I reworked manual unsafe managing of resources to Hotswap.

File: modules/rate-sampling/src/main/scala/io/janstenpickle/trace4cats/rate/sampling/RateSpanSampler.scala (hunk 1 above)
Comment (catostrophe) [reply modules/rate-sampling/src/main/scala/io/janstenpickle/trace4cats/rate/sampling/RateSpanSampler.scala:35]: Sure. Let's backport as much as we can if it's easy.

File: modules/dynamic-sampling/src/main/scala/io/janstenpickle/trace4cats/sampling/dynamic/SamplerProcess.scala [hunk 5]
```diff
@@ -0,0 +1 @@
+
```
Comment (catostrophe) [inline modules/dynamic-sampling/src/main/scala/io/janstenpickle/trace4cats/sampling/dynamic/SamplerProcess.scala:1]: This file should be deleted

**Comment by catostrophe** [comment]
I found a few mistakes in my yesterday's suggestions. I fixed them.

**Review by catostrophe — APPROVED** [review]

File: modules/dynamic-sampling-config/src/main/scala/io/janstenpickle/trace4cats/sampling/dynamic/config/ConfiguredHotSwappableDynamicSpanSampler.scala [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+import io.janstenpickle.trace4cats.kernel.SpanSampler
+import io.janstenpickle.trace4cats.model.{SampleDecision, SpanContext, SpanKind, TraceId}
+import io.janstenpickle.trace4cats.sampling.dynamic.HotSwappableDynamicSpanSampler
+
+trait ConfiguredHotSwappableDynamicSpanSampler[F[_]] extends SpanSampler[F] {
+  def updateConfig(config: SamplerConfig): F[Boolean]
+  def getConfig: F[SamplerConfig]
+}
+
+object ConfiguredHotSwappableDynamicSpanSampler {
```
Comment (janstenpickle) [inline modules/dynamic-sampling-config/src/main/scala/io/janstenpickle/trace4cats/sampling/dynamic/config/ConfiguredHotSwappableDynamicSpanSampler.scala:13]: I'm not sure this is the best name for this 😂

Comment (catostrophe) [reply modules/dynamic-sampling-config/src/main/scala/io/janstenpickle/trace4cats/sampling/dynamic/config/ConfiguredHotSwappableDynamicSpanSampler.scala:13]: @janstenpickle Looking at this I can taste good old Java in my mouth...

File: modules/dynamic-sampling/src/main/scala/io/janstenpickle/trace4cats/sampling/dynamic/HotSwappableDynamicSpanSampler.scala [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+  def create[F[_]: Temporal, A](
+    id: A,
+    initial: Resource[F, SpanSampler[F]]
+  ): Resource[F, HotSwappableDynamicSpanSampler[F, A]] = for {
+    (hotswap, sampler) <- Hotswap(initial)
+    current <- Resource.eval(Ref.of((sampler, id)))
+  } yield new HotSwappableDynamicSpanSampler[F, A] {
+    def updateSampler(newId: A, samplerResource: Resource[F, SpanSampler[F]]): F[Boolean] =
+      current.get
+        .map(_._2 != newId)
```
[the remark this answers is not in the document]
Comment (janstenpickle) [reply modules/dynamic-sampling/src/main/scala/io/janstenpickle/trace4cats/sampling/dynamic/HotSwappableDynamicSpanSampler.scala:26]: Oh yeah, let's go with `Eq` and maybe provide a default based on `equals`

File: modules/dynamic-sampling/src/main/scala/io/janstenpickle/trace4cats/sampling/dynamic/HotSwappableDynamicSpanSampler.scala [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+        .ifM(
+          hotswap.swap(samplerResource.evalTap(newSampler => current.set((newSampler, newId)))).as(true),
+          Applicative[F].pure(false)
+        )
```
Comment (catostrophe) [inline modules/dynamic-sampling/src/main/scala/io/janstenpickle/trace4cats/sampling/dynamic/HotSwappableDynamicSpanSampler.scala:30]:
```suggestion
        .flatTap { idChanged =>
          Applicative[F].whenA(idChanged)(
            hotswap.swap(samplerResource.evalTap(newSampler => current.set((newSampler, newId))))
          )
        }
```
Seems clearer, but don't insist

Comment (catostrophe) [reply modules/dynamic-sampling/src/main/scala/io/janstenpickle/trace4cats/sampling/dynamic/HotSwappableDynamicSpanSampler.scala]: I don't like ifM personally

File: modules/dynamic-sampling-config/src/main/scala/io/janstenpickle/trace4cats/sampling/dynamic/config/ConfiguredHotSwappableDynamicSpanSampler.scala (hunk 6 above)
Comment (janstenpickle) [reply modules/dynamic-sampling-config/src/main/scala/io/janstenpickle/trace4cats/sampling/dynamic/config/ConfiguredHotSwappableDynamicSpanSampler.scala:13]: I think it needs a few more `Proxy` and `Factory` in the name

File: modules/dynamic-sampling/src/test/scala/io/janstenpickle/trace4cats/sampling/dynamic/PollingDynamicSpanSamplerSpec.scala [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+    val test =
+      Ref.of[IO, (String, Resource[IO, SpanSampler[IO]])](("always", Resource.pure(SpanSampler.always))).flatMap {
+        samplerRef =>
+          PollingDynamicSpanSampler.create[IO, String](samplerRef.get, 1.second).use { sampler =>
+            val decision = sampler.shouldSample(None, TraceId.invalid, "test", SpanKind.Internal)
+
+            for {
+              d0 <- decision
+              _ <- samplerRef.set(("never", Resource.pure(SpanSampler.never)))
```
Comment (catostrophe) [inline modules/dynamic-sampling/src/test/scala/io/janstenpickle/trace4cats/sampling/dynamic/PollingDynamicSpanSamplerSpec.scala:36]: There was a typo

**Review by catostrophe — APPROVED** [review]

**Pull request merged**